fragment.js

```diff
@@ -134,3 +134,13 @@ function loadSettings(fragment) {
     }
     return settings
 }
+
+function loadFullSettings(fragment) {
```

Comment:
Is this necessary? A "complete" settings map isn't needed when loading the page initially. It should not be necessary when re-reading the fragment.

Reply:
The alternative is to check for each default and update it when moving back/forward. Not having either would mean the tab won't change when browsing (for example) from `tab=settings_tab` to `tab=totals_tab` (which would have no tab in the settings object).

Basically this function is just something that could be easily implemented as a one-off, but might be used in the future by other aspects of the code (I can't think of an example off the top of my head).

Personally I would have a global settings object and update it, then update the UI off the object (instead of referencing the UI when getting/changing settings), but I'll just do what you say because it's your code 😁

Reply:
Yeah, settings are all over the place. Some are global variables, some live in their UI elements, and some live on the FactorySpec. Refactoring them into an object dedicated for the purpose is on my to-do list, but I don't think this will be necessary here.

There's an invariant that is inconsistently applied across the settings. Enforcing it universally would make this simpler:

When rendering a setting, if its value is missing from the 'settings' object, it should be set to its default value.

Some of the render functions do this explicitly, while others do not, instead relying on some sort of global initialization to set the default. The lesson here is that global initialization is insufficient: renderSettings() might get called multiple times, and a settings object should *wholly* define the settings for the calculator, with the understanding that *missing* values should be equivalent to using the *default* values.

This should hold for all of the values in the fragment, actually, and not just those involved under renderSettings(). This is not a problem for "items" or "modules", which do this already, but as you point out, it is a problem for "tab". A solution there may be to have clickTab do e.g.

    if (!tabName) {
        tabName = DEFAULT_TAB
    }

Glancing through settings.js, I see that these settings do not appear to enforce this invariant:

* Display rate
* Rate and count precisions
* Furnace
* Mining productivity bonus
* Display format

The following *do* appear to enforce this invariant:

* Dataset
* Minimum assembler
* Preferred belt
* Minimum pipe length
* Default module
* Default beacon

If the settings in the first list can be corrected, this should remove the need to track the defaults explicitly.

Reply:
Cool, do you mind if I do that? Should it be included in this PR or get a separate one (I'm inclined to think include it because it's relevant)?

If I'm doing it then it _should_ be done in 24h, maybe 48h if I get busy.

Reply:
I was going to tweak this tonight. Hopefully it's as easy as I think it will be.

The default furnace is the "largest" furnace, as defined by FactoryDef.less(). Factories are ordered by the tuple of (crafting speed, module slots). In vanilla, this selects the electric furnace.

These criteria may change as part of adding Bob's support (depending on which furnace I want to use by default in Bob's), but I think the important thing is that the default is selected as part of initializing the FactorySpec (which parses through all of the factory entities). This should happen before the settings are rendered, so the default should be available by that point.

Reply:
Now I've got to learn how to merge changes 😛  Usually it's just me working on something